### PR TITLE
[WIP][examples/flax] add gradient accumulation

### PR DESCRIPTION
**Pull request by patil-suraj**
# What does this PR do?

Adds gradient accumulation in flax language modeling scripts.

### CODE REVIEW

File: examples/flax/language-modeling/run_clm_flax.py [hunk 1]
```diff
@@ -526,15 +537,28 @@ def compute_loss(params):
...
+            grads = jax.tree_map(lambda x: x / training_args.gradient_accumulation_steps, grad_accum)
+            grads = jax.lax.pmean(grads, "batch")
+            new_state = state.apply_gradients(
+                grads=grads, grad_accum=jax.tree_map(jnp.zeros_like, grads), optimizer_step=state.optimizer_step + 1
+            )
+            return new_state
 
-        new_state = state.apply_gradients(grads=grad, dropout_rng=new_dropout_rng)
+        new_state = jax.lax.cond(
+            state.step % training_args.gradient_accumulation_steps == 0,
```
Comment (patrickvonplaten) [inline examples/flax/language-modeling/run_clm_flax.py:552]: Does this make sure gradients are also accumulated for the first step? 
Is state.step = 1 the very first time?

File: examples/flax/language-modeling/run_clm_flax.py [hunk 2]
```diff
@@ -558,8 +582,11 @@ def eval_step(params, batch):
     logger.info(f"  Num examples = {len(train_dataset)}")
     logger.info(f"  Num Epochs = {num_epochs}")
     logger.info(f"  Instantaneous batch size per device = {training_args.per_device_train_batch_size}")
-    logger.info(f"  Total train batch size (w. parallel & distributed) = {train_batch_size}")
-    logger.info(f"  Total optimization steps = {total_train_steps}")
+    logger.info(
+        f"  Total train batch size (w. parallel & distributed) = {train_batch_size * training_args.gradient_accumulation_steps}"
+    )
+    logger.info(f"  Total global steps = {total_steps}")
+    logger.info(f"  Total optimization steps = {total_optimization_steps}")
```
Comment (patrickvonplaten) [inline examples/flax/language-modeling/run_clm_flax.py:589]: The line `576/603` `steps_per_epoch = len(train_dataset) // train_batch_size` should be updated no? `total_batch_size` should be used there and not `train_batch_size` no?

Comment (patrickvonplaten) [inline examples/flax/language-modeling/run_clm_flax.py:589]: We should probably also update

```cur_step = epoch * (len(train_dataset) // train_batch_size) + step``` in the line below to use `total_batch_size` no? 
Think when someone uses `gradient_accumulation` parameters like `logging_steps` should have have a different meaning as well no?

File: examples/flax/language-modeling/run_clm_flax.py [hunk 3]
```diff
@@ -169,6 +169,8 @@ def __post_init__(self):
 
 class TrainState(train_state.TrainState):
     dropout_rng: jnp.ndarray
+    grad_accum: jnp.ndarray
+    optimizer_step: int
```
Comment (patrickvonplaten) [inline examples/flax/language-modeling/run_clm_flax.py:173]: Would it maybe to just use `state.step` instead of creating a new `optimizer_step`?

File: examples/flax/language-modeling/run_clm_flax.py [hunk 4]
```diff
@@ -526,15 +537,28 @@ def compute_loss(params):
             return loss
 
         grad_fn = jax.value_and_grad(compute_loss)
-        loss, grad = grad_fn(state.params)
-        grad = jax.lax.pmean(grad, "batch")
+        loss, grads = grad_fn(state.params)
+        grad_accum = jax.tree_multimap(lambda x, y: x + y, grads, state.grad_accum)
+
+        def update_fn():
```
Comment (patrickvonplaten) [inline examples/flax/language-modeling/run_clm_flax.py:543]: We need better names here and more comments IMO or the script start becoming a bit unreadable

File: examples/flax/language-modeling/run_clm_flax.py [hunk 5]
```diff
@@ -526,15 +537,28 @@ def compute_loss(params):
...
+            grads = jax.tree_map(lambda x: x / training_args.gradient_accumulation_steps, grad_accum)
+            grads = jax.lax.pmean(grads, "batch")
+            new_state = state.apply_gradients(
+                grads=grads, grad_accum=jax.tree_map(jnp.zeros_like, grads), optimizer_step=state.optimizer_step + 1
```
Comment (patrickvonplaten) [inline examples/flax/language-modeling/run_clm_flax.py:547]: can't we just keep using `state.step` instead of passing a new variable `optimizer_step=state.optimizer_step + 1`?

File: examples/flax/language-modeling/run_clm_flax.py (hunk 1 above)
Comment (patrickvonplaten) [reply examples/flax/language-modeling/run_clm_flax.py:552]: I think this implies that the very first time this function is called that `state.step` is actually 0 which would mean that instead of accumulating the gradient an optimization step is done no?

**Comment by patrickvonplaten** [comment]
Thanks a lot for adding this! That's super useful! It seems to require some bigger changes to core functionality to the script so I think we should be careful here. Also I'm starting to wonder whether the examples become to complicated to read with more and more functionality being added and whether we should maybe instead creating a new training script instead?

Also wouldn't it be better to use gradient accumulation functionality from `optax` such as https://optax.readthedocs.io/en/latest/api.html?highlight=ApplyEvery#optax.apply_every ? Think a lot of the code that is written here already exists in optax classes/functions no?

@sgugger - I'd love to have your feedback on the PR as well

File: examples/flax/summarization/run_summarization_flax.py [hunk 6]
```diff
@@ -639,15 +650,28 @@ def compute_loss(params):
...
-        loss, grad = grad_fn(state.params)
-        grad = jax.lax.pmean(grad, "batch")
+        loss, grads = grad_fn(state.params)
+        grad_accum = jax.tree_multimap(lambda x, y: x + y, grads, state.grad_accum)
+
+        def update_fn():
+            grads = jax.tree_map(lambda x: x / training_args.gradient_accumulation_steps, grad_accum)
+            grads = jax.lax.pmean(grads, "batch")
+            new_state = state.apply_gradients(
+                grads=grads, grad_accum=jax.tree_map(jnp.zeros_like, grads), optimizer_step=state.optimizer_step
```
Comment (borisdayma) [inline examples/flax/summarization/run_summarization_flax.py:660]: I think the optimizer step is supposed to increase here

File: examples/flax/language-modeling/run_clm_flax.py (hunk 1 above)
Comment (sgugger) [reply examples/flax/language-modeling/run_clm_flax.py:552]: Yes it should probably be testing `state.step+1` against the `training_args.gradient_accumulation_steps`.

File: examples/flax/language-modeling/run_clm_flax.py [hunk 7]
```diff
@@ -558,8 +582,11 @@ def eval_step(params, batch):
     logger.info(f"  Num examples = {len(train_dataset)}")
     logger.info(f"  Num Epochs = {num_epochs}")
     logger.info(f"  Instantaneous batch size per device = {training_args.per_device_train_batch_size}")
-    logger.info(f"  Total train batch size (w. parallel & distributed) = {train_batch_size}")
-    logger.info(f"  Total optimization steps = {total_train_steps}")
+    logger.info(
+        f"  Total train batch size (w. parallel & distributed) = {train_batch_size * training_args.gradient_accumulation_steps}"
+    )
+    logger.info(f"  Total global steps = {total_steps}")
```
Comment (sgugger) [inline examples/flax/language-modeling/run_clm_flax.py:588]: The number of `total_steps` is the number of optimization steps. Our users are already confused so we need to be consistent with our convention across platforms.

File: examples/flax/language-modeling/run_clm_flax.py (hunk 2 above)
Comment (sgugger) [reply examples/flax/language-modeling/run_clm_flax.py:589]: Yes logging should be done every `logging_steps` full training steps, same for saves if you use `save_steps` in those scripts.

File: examples/flax/language-modeling/run_clm_flax.py [hunk 8]
```diff
@@ -461,14 +463,16 @@ def group_texts(examples):
     # Store some constant
     num_epochs = int(training_args.num_train_epochs)
     train_batch_size = int(training_args.per_device_train_batch_size) * jax.device_count()
+    total_batch_size = int(train_batch_size) * training_args.gradient_accumulation_steps
     eval_batch_size = int(training_args.per_device_eval_batch_size) * jax.device_count()
     steps_per_epoch = len(train_dataset) // train_batch_size
```
Comment (sgugger) [inline examples/flax/language-modeling/run_clm_flax.py:468]: This should be `total_batch_size` and not `train_batch_size`.

**Review by sgugger — COMMENTED** [review]
Thanks for the PR! It looks cool, but I have big rpoblems with the names :-) 

This does not follow the logic in naming we use in PyTorch or TensorFlow. A "step" is a training step, some step along the loss function following the gradients. When applying gradient accumulation, we do a step every `gradient_accumulation` training sample. I think the logs and the variable names should be adapted to reflect that, otherwise our users will be confused that we use one convention in PyTorch/TensorFlow and another one in Flax.

**Comment by stefan-it** [comment]
Hi, I've just checked the flax examples on master branch and it seems that gradient accumulation is still missing, so I'm coming back to this PR :)

@patrickvonplaten mentioned to use `optax`, and I've found this (working?) implementation of gradient acc. for T5 MLM pre-training from @gsarti. This may could help here :hugs:

**Comment by ArthurZucker** [comment]
Closing as this PR is super old
